Introduce themecolor

benwebber opened the pull request:
Changes:

* Adds global `trimcolor`, which colours `commentbox`, `dndtable`, and `paperbox`.
* Adds helper commands for setting/resetting `trimcolor`.
* Adds trim colours from all parts of the PHB and DMG (sampled from scans).
* `paperbox` uses the same colour as `commentbox` and `dndtable`.
* `dnditemtable` wraps `dndtable`.

Colour names use PascalCase because they are very hard to read otherwise. X11 uses the same convention.

Resolves #48.

BrianCriswell commented:
It is looking great! A few^H^H^H lot of thoughts:
1. When I envisioned trimcolor, the idea was that it would be a quick way for a user to keep the color of the commentbox, paperbox, and table synchronized. However, a user should still have the option to have different colors for each element. So commentboxcolor, paperboxcolor, and tablecolor are still needed and should be what are used in the various .sty files. Those colors should be initialized to trimcolor. A user can change trimcolor to change all three or change each one individually. Originally, I thought that a user could change trimcolor and have that change the other three, but I just realized that is not how LaTeX operates. This leads to:
2. I know I originally suggested trimcolor be used to control the colors commentboxcolor, paperboxcolor, and tablecolor, but I think it would be more useful as a hook into the current themecolor  for the trim elements a user adds to the document that they want synced with the built in elements. 
3. Building off of 2, maybe a fifth color should be added called themecolor, and that is used to set the other four main colors. `\setthemecolor{color}` would set themecolor and the other four colors, and `\setthemecolor` without the optional color parameter would set the four main colors back to themecolor.
4. In lib/dndcolor.sty it should be `\colorlet` not `\Colorlet`.
5. Pascal case for acronyms has every character of two character acronyms capitalized and only the first character of a longer acronym capitalized. Examples are PhbBlue, MMBlue, and DmgBlue.
6. Using an HTML color code with `\definecolor` requires capital letters.
7. I do not think we need both the book colors named for their color and their usage. I would lean toward the color (i.e. PhbLightGreen).
8. I would eventually like to get the other colors like titlered and undergold more generalized as well, but we can tackle those at a later date.
9. Similarly to 8, I would like to give hooks to tie into the colors of the text portion of the themed elements at a later date (for the times when themecolor is too dark for black text).
10. Can you give an example of the tabularx problem that caused you to rewrite dndtable? I am surprised you are having trouble with it as the way it was written was essentially a copy of the example from the tabularx manual.

Thanks for all your work on this, Ben.

benwebber commented:
Added four element colours and `themecolor`.

---

> 4. In lib/dndcolor.sty it should be `\colorlet` not `\Colorlet`.

Oops, fixed.

> 5. Pascal case for acronyms has every character of two character acronyms capitalized and only the first character of a longer acronym capitalized. Examples are PhbBlue, MMBlue, and DmgBlue.

Agree, changed.

> 6. Using an HTML color code with `\definecolor` requires capital letters.

It worked with the lowercase hex values, but corrected anyway.

> 7. I do not think we need both the book colors named for their color and their usage. I would lean toward the color (i.e. PhbLightGreen).

Agree, changed.

> 10. Can you give an example of the tabularx problem that caused you to rewrite dndtable? I am surprised you are having trouble with it as the way it was written was essentially a copy of the example from the tabularx manual.

It's specifically for environments that wrap an environment that wrap tabularx. This causes a runaway argument error:

```tex
\newenvironment{dnditemtable}[1]
  {%
    \colorlet{oldtrimcolor}{trimcolor}
    \settrimcolor{DmgCoral}
    \begin{dndtable}{#1}
  }
  {%
    \end{dndtable}
    \settrimcolor{oldtrimcolor}
  }
```

The tabularx manual suggests grouping the "contents" of a tabularx. That resolved the runaway argument error, but left `dnditemtable` unstyled.

BrianCriswell commented on lib/dndcomment.sty:
```diff
@@ -14,19 +14,9 @@
-	colback=commentgreen,
-	colframe=commentgreen,
-	colbacktitle=commentgreen,
+	colback=commentboxcolor,
```
Can you change the definition of commentbox to match the definition of paperbox (i.e. with the optional color)?

benwebber replied:
Done.

BrianCriswell commented on lib/dndtable.sty:
```diff
@@ -9,11 +9,31 @@
+% Wrapping an environment that uses tabularx causes a runaway argument error
+% because tabularx cannot find the end token.
+% Instead, create two command that expand to the table header and footer.
+\newcommand{\dnd@starttable}[1]{%
```
I do not think we need this, but I am not completely against it. dnditemtable is deprecated by this change, and I cannot think of other reasons to nest the dndtable environment (doesn't mean there aren't any). This is part of the tradeoff of the easier column layout of tabularx I guess.

BrianCriswell commented on lib/dndtable.sty:
```diff
@@ -9,11 +9,31 @@
+
+\newenvironment{dndtable}[1][XX]{\dnd@starttable{#1}}{\dnd@endtable}
+
+\newenvironment{dnditemtable}[1][XX]
```
dnditemtable is deprecated by this change and should probably not be changed to fit into the new color method of handling colors. If you are feeling adventurous, you could try adding this to the beginning of the original dnditemtable definition:
```
\PackageWarning{dnd}{dnditemtable is deprecated. use dndtable after setting tablecolor instead.}
```
You could also change the definition of the dndtable environment to take an optional color like paperbox does.

benwebber replied:
Done.

BrianCriswell commented on lib/dndcolors.sty:
```diff
@@ -0,0 +1,61 @@
+  \colorlet{themecolor}{#1}
+  \colorlet{trimcolor}{#1}
+}
+\newcommand{\resetthemecolor}{\setthemecolor[PhbLightGreen]}
```
I still think we should probably skip `\resetthemecolor`.

benwebber replied:
On retrospect it might be confusing (i.e., does it set the default colour or the previous colour after changing it?). Removed.

BrianCriswell commented on lib/dndcolors.sty:
```diff
@@ -0,0 +1,61 @@
+% Type
+\definecolor{titlered}{HTML}{58180D}	% titles
+\definecolor{titlegold}{HTML}{C9AD6A} % titlerules
+\definecolor{rulered}{HTML}{9C2B1B}		% triangular rule in statsblock
```
These colors that specifically mention an actual color in their name could be changed now to more generic names:
```
\definecolor{DndTitleRed}{HTML}{58180D}
\colorlet{titlecolor}{DndTitleRed}
```
Although you have already done a lot of work on this, so we could push it off until later. If you are feeling "done" with this.

benwebber replied:
While those colours should be "public" (i.e., documented), they don't change with the theme. I'd prefer to keep this PR scope limited to the theme colours.
